### PR TITLE
Add new framework for testing

### PR DESCRIPTION
**Pull request by joemphilips**
This will launch new docker-compose instance for each test.
Which enables more flexible testing.

### CODE REVIEW

**Comment by joemphilips** [comment]
Now it works fine. But with some nits.
1. It is slow.
1. When I finish test with Ctrl-C , `Dispose()` will not run as I expect. So containers must be stopped manually. 
 1. It puts all data created by nodes in specific directory under `bin/` (It is same with how `NBitcoin.TestFramework` manages data directory.) But clightning did not work with bind mounts, so it uses data volumes instead.

File: tests/docker-compose.flexible.yml [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      - "9835" # api port
+    volumes:
+      - "${TESTER_DATADIR:?NoTesterDataDir}/bitcoind_dir:/etc/bitcoin"
+      - "lightningd_dest_dir:/root/.lightning"
+    links:
+      - bitcoind
+
+  lnd:
+    restart: unless-stopped
+    image: btcpayserver/lnd:v0.5.1-beta-2
```
Comment (NicolasDorier) [inline tests/docker-compose.flexible.yml]: 0.5.2 now

**Comment by NicolasDorier** [comment]
I am not against, but I remember struggling a big time to make it reliable in the past.

See https://github.com/dgarage/LightningBenchmarks/tree/master/bench/Lightning.Bench , I did kind of similar thing.

Basically, I made a framework which:

1. Generate a docker compose dynamically
2. Run it
3. Run the benchmarks
4. Tear it down

I think to make it more useful, your builder should:

1. Be in a separate project library (in the src folder) so we can make a nuget package later
2. Dynamically create the compose files (via the builder) instead of having `docker-compose.flexible.yml`.

**Comment by joemphilips** [comment]
I see, I will try. But it may take some time.

File: tests/CommonTests.cs [hunk 2]
```diff
@@ -155,6 +160,74 @@ public async Task CanListChannels()
             }
         }
 
+        [Fact]
+        public async Task CanListChannelsInAllCase()
+        {
+            var builder = new FlexibleTesterBuilder();
```
Comment (joemphilips) [inline tests/CommonTests.cs]: See here for how to use.

**Comment by joemphilips** [comment]
I'm not sure it is the best way. But now it works fine.

**Comment by NicolasDorier** [comment]
Cool stuff! I am fine for merging it but I am wondering if that's the best.

I will probably not dog food it for quite a while, while I think you already want to use it or already is using it.

In this case, why don't you create a new repository so I don't slow you down?

I see a reason to keep it in this repository though: So we use it in in the tests of this repo.

Both is fine for me. What do you think?

**Comment by joemphilips** [comment]
Well the thing is, I originally started this project in my own repo here (https://github.com/joemphilips/LNTestFramework)
If you think it is fine to have this library as external dependency, I will complete it as my own project (probably in F#), and reference it from this library.

But please support SourceLink by the way, it makes interoperation much much easier.

**Comment by NicolasDorier** [comment]
@joemphilips an idea is to add your library as a submodule in the repo. That we can use your test framework in this repo.

BTW, since I updated to 0.7.0, the `connect` from c-lightning to lnd is freezing. :(

**Comment by NicolasDorier** [comment]
I just added symbol (snpkg) so you can step in sources in the new version.

**Comment by joemphilips** [comment]
Cool. Thanks.
I will come back to this after finishing miniscript.

**Comment by rockstardev** [comment]
@joemphilips I'll close this PR since a lot has changed in a year - feel free to reopen new PR if you want to revisit this. <3

**Pull request closed without merging**